From b2f536b60ac18a2e7059b3aba7a0ffb23213146e Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:37:54 -0400
Subject: [virt] kvm/nept: reset PDPTR register cache on nested vmentry emulation

Message-id: <1406040016-3289-80-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86158
O-Subject: [RHEL7 PATCH v2 079/221] KVM: nEPT: reset PDPTR register cache on nested vmentry emulation
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Gleb Natapov <gleb@redhat.com>

After nested vmentry stale cache can be used to reload L2 PDPTR pointers
which will cause L2 guest to fail. Fix it by invalidating cache on nested
vmentry emulation.

https://bugzilla.kernel.org/show_bug.cgi?id=60830

Signed-off-by: Gleb Natapov <gleb@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 72f857950f6f19cba42a9ded078bbc99f10aa667)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 963b866..708ecc3 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -7776,6 +7776,10 @@ static void prepare_vmcs02(struct kvm_vcpu *vcpu, struct vmcs12 *vmcs12)
   vmcs_write64(GUEST_PDPTR1, vmcs12->guest_pdptr1);
   vmcs_write64(GUEST_PDPTR2, vmcs12->guest_pdptr2);
   vmcs_write64(GUEST_PDPTR3, vmcs12->guest_pdptr3);
+  __clear_bit(VCPU_EXREG_PDPTR,
+    (unsigned long *)&vcpu->arch.regs_avail);
+  __clear_bit(VCPU_EXREG_PDPTR,
+    (unsigned long *)&vcpu->arch.regs_dirty);
  }
 
  kvm_register_write(vcpu, VCPU_REGS_RSP, vmcs12->guest_rsp);
-- 
1.7.1